feat: usage/record capability definition

fforbeck opened the pull request:
### Add `usage/record` Capability definition for Egress Traffic Tracking

**Summary:**
- Introduced a new capability `usage/record` definition to track egress traffic metrics.
- The `w3upInfra` upload-api will be updated to implement this new capability by leveraging Stripe's Usage Records API for accurate tracking and billing.

RFC: https://github.com/storacha/RFC/pull/37

alanshaw commented on packages/capabilities/src/types.ts:
```diff
@@ -161,6 +165,17 @@ export interface UsageData {
+  /** Id of the customer that is being billed. */
+  customer: AccountDID
+  /** CID of the resource that was served. */
+  resourceCID: string
```
```suggestion
  resource: Link
```

alanshaw commented on packages/capabilities/src/usage.js:
```diff
@@ -40,3 +40,22 @@ export const report = capability({
+    /** MailTo DID of the customer that is being billed. */
+    customer: AccountDID,
+    /** CID of the resource that was served. */
+    resourceCID: Schema.string(),
```
```suggestion
    resource: Schema.link(),
```

alanshaw commented on packages/upload-api/test/storage/usage-storage.js:
```diff
@@ -64,4 +68,29 @@ export class UsageStorage {
+   * @param {number} bytes
+   * @param {Date} servedAt
+   */
+  async record(customer, resourceCID, bytes, servedAt) {
```
Please lets not specify types in variable names.

alanshaw commented on packages/w3up-client/src/capability/usage.js:
```diff
@@ -31,6 +31,32 @@ export class UsageClient extends Base {
+      { agent: this.agent },
+      { ...options, space, egressData }
+    )
+    /* c8 ignore next 7 */
```
7 includes the return value

alanshaw commented on packages/upload-api/src/usage/record.js:
```diff
@@ -0,0 +1,24 @@
+const record = async ({ capability }, context) => {
+  const res = await context.usageStorage.record(
+    capability.nb.customer,
+    capability.nb.resourceCID,
```
Use `UnkonwnLink` for links you don't know the CID version of.

fforbeck commented on packages/capabilities/src/usage.js:
```diff
@@ -40,3 +40,22 @@ export const report = capability({
+  with: SpaceDID,
+  nb: Schema.struct({
+    /** MailTo DID of the customer that is being billed. */
+    customer: AccountDID,
```
Yeah, I've discussed this with @Peeja and it will be available in Freeway at the time the invocation is triggered, here is the convo: https://github.com/storacha/RFC/pull/36#discussion_r1790818236

fforbeck replied:
@travis The Provider DID returned by that function is `did:web:string`, but the customer DID is type `AccountID` = `did:mailto:...`. Any suggestions on how to convert it to a mailto DID?

fforbeck commented on packages/upload-api/src/usage/record.js:
```diff
@@ -0,0 +1,24 @@
+ * @returns {Promise<API.Result<API.EgressRecordSuccess, API.EgressRecordFailure>>}
+ */
+const record = async ({ capability }, context) => {
+  const res = await context.usageStorage.record(
```
We will have the customer ID provided during the invocation in Freeway. We could do it here in another moment, but I think that for the first iteration, it should be easy enough to get that in Freeway considering the changes that @Peeja will be working on.

alanshaw replied:
Correct, no account in location commitment, just space.

fforbeck commented on packages/w3up-client/src/capability/usage.js:
```diff
@@ -31,6 +31,32 @@ export class UsageClient extends Base {
+      { agent: this.agent },
+      { ...options, space, egressData }
+    )
+    /* c8 ignore next 5 */
```
Sounds good. I will add that.

fforbeck replied:
@travis I tried to mock the `agent.invokeAndExecute` to return an error using Sinon, but for some reason, it doesn't work. I guess that's why the other capability also had the `c8 ignore` for this part.

fforbeck commented on packages/upload-api/src/usage/record.js:
```diff
@@ -0,0 +1,40 @@
+ * @returns {Promise<API.Result<API.EgressRecordSuccess, API.EgressRecordFailure>>}
+ */
+const record = async ({ capability, invocation }, context) => {
+  const provider = /** @type {`did:web:${string}`} */ (
+    invocation.audience.did()
+  )
+  const consumerResponse = await context.provisionsStorage.getConsumer(
+    provider,
+    capability.with
+  )
+  if (consumerResponse.error) {
+    return {
+      error: {
+        name: 'EgressRecordFailure',
+        message: `Failed to get consumer`,
+      },
+    }
+  }
+  const consumer = consumerResponse.ok
+  const res = await context.usageStorage.record(
+    consumer.customer,
+    capability.nb.resource,
+    capability.nb.bytes,
+    new Date(capability.nb.servedAt * 1000)
```
Updated the function to find the `customer` using the provider `invocation.audience` and consumer (`capability.with`), the `customer` represent the account we want to charge for the egress traffic.

fforbeck commented on packages/capabilities/src/types.ts:
```diff
@@ -193,6 +208,7 @@ export interface ConsumerGetSuccess {
   allocated: number
   limit: number
   subscription: string
+  customer: AccountDID
```
`Customer` is now part of the `ConsumerGetSuccess` response.

alanshaw commented on packages/upload-api/src/usage/record.js:
```diff
@@ -0,0 +1,40 @@
+    return {
+      error: {
+        name: 'EgressRecordFailure',
+        message: `Failed to get consumer`,
```
```suggestion
        message: 'Failed to get consumer',
```

alanshaw commented on packages/upload-api/src/usage/record.js:
```diff
@@ -0,0 +1,40 @@
+    capability.with
+  )
+  if (consumerResponse.error) {
+    return {
```
This actually returns `ConsumerNotFound` error when there's no subscription setup. It's arguably more useful than "Failed to get consumer".

```suggestion
    return consumerResponse
```

Really we need to be better at defining error types, as the failure type for `getConsumer` is currently the catchall `Failure`.

If you're up for it, it would be good to add a type for `ConsumerNotFound` and include it as a possible value for `EgressRecordFailure`.

i.e.
```ts
type EgressRecordFailure = ConsumerNotFound | Ucanto.Failure
```

fforbeck replied:
That makes sense to me. I will update the types. Thanks

alanshaw commented on packages/upload-api/src/types/usage.ts:
```diff
@@ -13,4 +15,10 @@ export interface UsageStorage {
+    customer: AccountDID,
+    resource: UnknownLink,
+    bytes: number,
+    servedAt: Date
```
We should add a `cause: Link` field to identify the invocation.

fforbeck replied:
@alanshaw Where does it come from?

alanshaw replied:
In a ucanto handler it is `invocation.cid`

alanshaw commented on packages/upload-api/src/types/usage.ts:
```diff
@@ -13,4 +15,10 @@ export interface UsageStorage {
   ) => Promise<Result<UsageData, Failure>>
+  record: (
+    customer: AccountDID,
+    resource: UnknownLink,
```
Would be good to get a comment here - what is the resource? It's the root CID of a DAG (not the CID of a CAR) or I imagine it could also be literally a CID of any sub-DAG within a DAG.

fforbeck replied:
It is the same data CID we use in the gateway to fetch the content. It is pulled from the URL via `withParsedIpfsUrl` handler from the `gateway-lib`: https://github.com/storacha/gateway-lib/blob/main/src/middleware.js#L118
I think this is the Root CID of a DAG, but I am not 100% sure. Do you know?

alanshaw replied:
https://w3s.link/ipfs/bafybeibrqc2se2p3k4kfdwg7deigdggamlumemkiggrnqw3edrjosqhvnm/yesthisisdog.jpg

In the above URL, the root CID is `bafybeibrqc2se2p3k4kfdwg7deigdggamlumemkiggrnqw3edrjosqhvnm`.

If you look at the HTTP headers, the etag will tell you the CID of the resource at the end of the path i.e. the CID of `yesthisisdog.jpg`, which is `bafkreiem4twkqzsq2aj4shbycd4yvoj2cx72vezicletlhi7dijjciqpui`.

You can use this as a root CID too!
https://w3s.link/ipfs/bafkreiem4twkqzsq2aj4shbycd4yvoj2cx72vezicletlhi7dijjciqpui

If you look at the HTTP headers, the etag is `bafkreiem4twkqzsq2aj4shbycd4yvoj2cx72vezicletlhi7dijjciqpui`, since there is no path.

So when I say "root" CID I mean the CID that might be considered the root CID of a given upload. i.e. a CID that would be set as `root` in an `upload/add` invocation.

All I think is needed is a clarification that the `resource` as you have it there is _not_ the CID of a blob/shard, it's the CID of some gateway accessible content.

alanshaw commented on packages/upload-api/src/types/usage.ts:
```diff
@@ -13,4 +15,10 @@ export interface UsageStorage {
   ) => Promise<Result<UsageData, Failure>>
+  record: (
+    customer: AccountDID,
+    resource: UnknownLink,
```
Do we have enough information to identify the blob(s)/shard(s) the resource was served from? Feels like a useful thing to track here, especially if the CID being used is not a root CID, but a sub-DAG. I'm thinking about users querying why a certain CID has used so much egress but it is not one of the root CIDs from `upload/add` - how do we prove it belongs to them?

alanshaw commented on packages/upload-api/src/usage/record.js:
```diff
@@ -0,0 +1,40 @@
+    capability.nb.bytes,
+    new Date(capability.nb.servedAt * 1000)
+  )
+  if (res.error) return res
+
+  return res
```
```suggestion
  return res
```

alanshaw commented on packages/w3up-client/src/capability/usage.js:
```diff
@@ -61,3 +87,32 @@ export const report = async (
+    proofs,
+    nonce,
+    nb: {
+      customer: egressData.customer,
```
```suggestion
```

alanshaw commented on packages/capabilities/src/types.ts:
```diff
@@ -161,6 +165,17 @@ export interface UsageData {
 }
 
+export interface EgressData {
+  /** Id of the customer that is being billed. */
+  customer: AccountDID
```
The service needs to obtain this.

```suggestion
```

alanshaw commented on packages/capabilities/src/usage.js:
```diff
@@ -40,3 +40,22 @@ export const report = capability({
+  can: 'usage/record',
+  with: SpaceDID,
+  nb: Schema.struct({
+    /** MailTo DID of the customer that is being billed. */
+    customer: AccountDID,
```
```suggestion
```

alanshaw commented on packages/upload-api/src/types/usage.ts:
```diff
@@ -13,4 +15,10 @@ export interface UsageStorage {
     period: { from: Date; to: Date }
   ) => Promise<Result<UsageData, Failure>>
+  record: (
+    customer: AccountDID,
```
Record space DID also? Would be good to be able to enumerate usage by space.

alanshaw commented on packages/w3up-client/test/capability/usage.test.js:
```diff
@@ -68,6 +69,47 @@ export const UsageClient = Test.withContext({
+      assert.ok(result)
+
+      const record = await alice.capability.usage.record(space.did(), {
+        customer: 'did:mailto:alice@web.mail',
```
```suggestion
```

fforbeck commented:
@alanshaw I've implemented the suggested changes and added an additional note to the `resource` field explaining that it is not a CID of a blob/shard. It is ready for another review when you get a chance.

fforbeck commented:
@travis I've implemented the suggested changes. It is ready for another review.

alanshaw commented on packages/w3up-client/src/capability/usage.js:
```diff
@@ -61,3 +92,35 @@ export const report = async (
+ * @param {API.SpaceDID} egressData.space
+ * @param {API.UnknownLink} egressData.resource
+ * @param {number} egressData.bytes
+ * @param {string} egressData.servedAt
```
You could make this optional and default to now.
